ActiveStorage Guide

jeffreyguenther opened the pull request:
### Summary

This PR is meant to be a placeholder for the ActiveStorage guide.

Having implemented a service for [OpenStack](https://github.com/jeffreyguenther/activestorage-openstack), I have a sense of what's necessary to build a new service and to integrate ActiveStorage into an app.

This PR is a work in progress and I hope to gather feedback as the guide comes together.

Topics:
- [x] How to attach a file(s) to a model.
- [x] How to remove the attached file.
- [x] How to link to the attached file.
- [x] How to implement a download link.
- [x] How to create variations of an image.
- [x] How to generate a preview for files other than images.
- [x] How to upload files directly to a service.
- [x] How to add support for additional cloud services.
- [x] How to clean up files during testing - use lessons learned in #30935 
- [x] Example how to use direct upload events in a form

jeffreyguenther committed:
Begins guide for ActiveStorage

jeffreyguenther committed:
Fixes typo in docs

jeffreyguenther committed:
Adds todo for preview

jeffreyguenther committed:
Adds content from README and conversations with George

rafaelfranca commented:
r? @georgeclaghorn

jeffreyguenther committed:
Merge branch 'master' into activestorage-guide

georgeclaghorn commented:
Thanks for kicking this off, @jeffreyguenther! :heart: I’ll give a thorough review before the end of the week.

jeffreyguenther committed:
Adds more explanation to points and add setup section

jeffreyguenther committed:
Adds testing section

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,305 @@
+    root: <%= Rails.root.join("storage") %>
+```
+NOTE: Should we include the required keys for all the supported services?
+NOTE: Should we mention the mirror service and how to set it up?
````
Yes and yes.

georgeclaghorn replied:
To expand on my answer to the first question, let’s show the minimum required config for each service and refer readers to client library docs for optional config.

georgeclaghorn commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,305 @@
+-------------------------------
+
+To remove an attachment from a model, call `purge` on the attachment. Removal
+can be done in the background if your application is setup to use ActiveJob.
```
Setup is a noun. The verb form is “set up.”

Use “Active Job” (with a space between the words) to refer to the framework.

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,305 @@
+
+# Destroy the associated models and actual resource files async, via Active Job.
+user.avatar.purge_later
+```
````
Let’s be clear that purging an attachment deletes its underlying blob and removes the stored file from the service.

georgeclaghorn commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,305 @@
+access, a redirect to the actual service endpoint is returned. This indirection
+decouples the public URL from the actual one, and allows for example mirroring
+attachments in different services for high-availability. The redirection has an
+HTTP expiration of 5 min.
```
> This indirection decouples the public URL from the actual one, and allows for example mirroring attachments in different services for high-availability.

The primary reason for pointing to a stable application URL is to support page and fragment caching. A short-lived URL like one returned by `ActiveStorage::Service#service_url` can’t be cached. I’d just say that: “decoupling” is too abstract a benefit for a usage guide, and mirroring isn’t (as far as I can tell) a good justification for the redirect.

georgeclaghorn reviewed (changes requested):
Here’s a first pass. I have more to add later. Great start!

jeffreyguenther commented:
Thanks @georgeclaghorn! I'll work on these!

jeffreyguenther commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,202 @@
+* How to create variations of an image.
+* How to generate a preview for files other than images.
+* How to upload files directly to a service.
+* How to implement a download link.
```
Are you thinking that "How to link to an attached file" should include a description how to create a download link with an attachment disposition?

georgeclaghorn replied:
Yeah, I think we can explain the `disposition:` option when we show how to generate a URL.

jeffreyguenther committed:
Integrates George's suggestions

- Describes how to setup each of the services in the `services.yml`
- Integrates copy changes

jeffreyguenther commented:
I haven't addressed the download link yet as I'm a little fuzzy on how to do this. `url_for`, as far as I know, doesn't allow you to pass it params when you're passing it an object. Is that correct?

When I recently implemented a download link, I created a route + controller to handle the download and it passed the disposition into `service_url`. Have I missed how this should be done?

jeffreyguenther committed:
Proofreading

simonista commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,388 @@
+  bucket: ""
+```
+
+Also, add the S3 client gem to your Gemfile:
````
s/S3/Google Cloud Storage/

jeffreyguenther replied:
Good catch! Copy and paste error.

georgeclaghorn commented:
@jeffreyguenther That sounds correct. We can recommend `rails_blob_url` when params need to be provided.

pacMakaveli commented:
Hi guys.

Considering most of the Rails community has been using Paperclip ( mainly ) to add attachments in their app, will there ever be a migration guide from Paperclip to ActiveStorage? Or maybe, even better imo, does ActiveJob provide an option to customize the S3 path?
Is this even something you guys would cover!?

Thanks!

jeffreyguenther commented:
@pacMakaveli I maintain a couple apps that use Paperclip myself so I have given some thought to the problem. I don’t have any plans to write anything to that end. In my view, it’s outside the scope of this guide. 

Paperclip and Active Storage have quite different approaches to the way files are organized on the cloud service, so the process will be involved, even more so if you try to preserve your formats as variants. Personally, I wouldn’t do that. I would only move the originals and let Active Storage generate the variants on demand as your application is used.

jeffreyguenther committed:
Adds download link helper explanation

jeffreyguenther committed:
Merge branch 'master' into activestorage-guide

jeffreyguenther commented:
@georgeclaghorn Should it be possible for users to set the filename of the file downloaded?

[BlobsController](https://github.com/rails/rails/blob/master/activestorage/app/controllers/active_storage/blobs_controller.rb#L11) currently only passes the disposition to the `service_url` and uses the Blob's file name as the download file name. `Service#url` does take a filename so it is possible with a small change.

jeffreyguenther committed:
Updates direct download discussion

jeffreyguenther commented:
@georgeclaghorn Also, I'm not sure that `rails_blob_url` works for downloads. I'm not familiar with how these internal routing methods work, but I don't think the disposition is passed through.

This doesn't work:
``` ruby 
rails_blob_url(audio_blob, disposition: "attachment")
```

georgeclaghorn commented:
No, the blob’s filename should always be used. 

The other thing is a bug. [`rails_blob_url`](https://github.com/rails/rails/blob/master/activestorage/config/routes.rb#L6-L8) should pass any options it receives to `rails_service_blob_url`:

```diff
- direct :rails_blob do |blob|
-   route_for(:rails_service_blob, blob.signed_id, blob.filename)
+ direct :rails_blob do |blob, options = {}|
+   route_for(:rails_service_blob, blob.signed_id, blob.filename, options)
  end
```

jeffreyguenther committed:
Remove mention of filename from download

jeffreyguenther committed:
Adds tools caveat for generating previews

georgeclaghorn commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,396 @@
+
+  def create
+    message = Message.create! params.require(:message).permit(:title, :content)
+    message.images.attach(params[:message][:images])
```
As of 1d24e47, images can be attached in the `create!` call:

```ruby
def create
  message = Message.create!(message_params)
  redirect_to message
end

private
  def message_params
    params.require(:message).permit(:title, :content, images: [])
  end
```

jeffreyguenther committed:
Updates add multiple attachments example

jeffreyguenther committed:
Merge branch 'master' into activestorage-guide

jeffreyguenther committed:
Corrects Active Job spelling

jeffreyguenther commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,400 @@
+| `direct-upload:end` | `<input>` | `{id, file}` | A direct upload has ended. |
+| `direct-uploads:end` | `<form>` | None | All direct uploads have ended. |
+
+NOTE: Is there more to using the direct upload than this? How does one associate
```
@georgeclaghorn Do you know the answer or where I can look to find it?

georgeclaghorn replied:
> Is there more to using the direct upload than this?

Nope. The direct upload JS inserts a hidden input with the same name as the file input and sets its value to the directly-uploaded blob's signed ID. The signed blob ID is attachable: it can be passed to `attach` or assigned to the attachment in `new`/`create`.

jeffreyguenther replied:
Awesome! So nice that these details are handled for us. With other tools, we'd before to handle this on our own.

jeffreyguenther committed:
Removes note

jeffreyguenther commented:
I think we're ready for a thorough proofreading/formatting pass. I don't have anything else to add to the guide. Is there anything else you'd like to see added?

javan commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,397 @@
+    ```
+3. That's it! Uploads begin upon form submission.
+
+### Direct upload JavaScript events
````
Here's an example of how you might use these events in an app to display direct upload progress: https://gist.github.com/javan/5538692cb37a683db15792be8d05761e. Feel free to borrow (or ignore)!

jeffreyguenther replied:
Thanks! Good idea! I'll incorporate your gist. I find the more tangible examples we can provide the better people will understand what's possible.

georgeclaghorn commented:
Fixed the route helper bug in https://github.com/rails/rails/commit/3fa812615a28f9c6392c433f3b08c41c5efb999f.

jeffreyguenther committed:
Merge branch 'master' into activestorage-guide

tradeli commented:
Per this draft, it looks like ActiveStorage is only useful with non-API Rails apps because of the Active View integration, right?

rafaelfranca commented:
@tradeli no. Of course you will not take fully advantage of Active Strorage, but you can use it without Action View.

georgeclaghorn commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,397 @@
+
+Inside a Rails application, you can set up your services through the generated
+`config/storage.yml` file and reference one of the supported service types under
+the `service` key.
```
Elaborate on the structure of `config/storage.yml` and provide an example:

> Declare Active Storage services in `config/storage.yml`. For each service your application uses, provide a name and the requisite configuration. The example below declares three services named `local`, `test`, and `s3`:
>
>```yaml
> local:
>   service: Disk
>   root: <%= Rails.root.join("storage") %>
>
> test:
>   service: Disk
>   root: <%= Rails.root.join("tmp/storage") %>
>
> s3:
>   service: S3
>   access_key_id: ""
>   secret_access_key: ""
> ```
>
> Continue reading for information on the built-in service adapters (e.g. `Disk` and `S3`) and the configuration they require.

Expand on how and when to specify which service Active Storage should use:

> Tell Active Storage which service to use by setting `Rails.application.config.active_storage.service`. Because each environment will likely use a different service, it is recommended to do this on a per-environment basis. To use the `disk` service from the previous example in the development environment, you would add the following to `config/environments/development.rb`:
>
> ```ruby
> # Store files locally.
> config.active_storage.service = :disk
> ```
>
> To use the `s3` service in production, you would add the following to `config/environments/production.rb`:
>
> ```ruby
> # Store files in S3.
> config.active_storage.service = :s3
> ```

Because populating `config/storage.yml` and setting `Rails.application.config.active_storage.service` are required and using image variants is optional, I would put the information on adding `mini_magick` to the Gemfile last.

georgeclaghorn commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,397 @@
+the `service` key.
+
+### Disk Service
+To use the Disk service:
```
Let’s clarify that the Disk service is intended for local use only. It shouldn’t be used in production.

jeffreyguenther replied:
What about for small apps that don't warrant setting up cloud storage? Other gems like Paperclip allow local storage. Are there performance reasons why this shouldn't be done?

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,397 @@
+
+To use Amazon S3:
+``` yaml
+local:
````
Give this a different name. `local` isn’t quite right.

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,397 @@
+To use Microsoft Azure Storage:
+
+``` yaml
+local:
````
Give this a different name. `local` isn’t quite right.

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,397 @@
+To use Google Cloud Storage:
+
+``` yaml
+local:
````
Give this a different name. `local` isn’t quite right.

georgeclaghorn commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,397 @@
+You can keep multiple services in sync by defining a mirror service. When
+a file is uploaded or deleted, it's done across all the mirrored services.
+Define each of the services you'd like to use as described above and then define
+a mirrored service which references them.
```
Explain that files are always served from the primary service. Only writes are mirrored.

Comment on why you’d use the Mirror service. It’s useful in carrying out a migration between services in production. You can start mirroring to the new service, copy existing files from the old service to the new, then go all-in on the new service.

jeffreyguenther replied:
Doesn't 

https://github.com/rails/rails/blob/f7e3c686685fb89e67293440d24356f93fa34847/activestorage/lib/active_storage/service/mirror_service.rb#L34-L36

mean that deletes are done across all services?

georgeclaghorn replied:
A delete is a write. :smile:

jeffreyguenther replied:
Touché! 😁

jeffreyguenther commented:
With DHH's announcement of the 5.2 beta, it looks like we're getting close to a release. I'll do my best to get these edits done in the next day. Hate to be blocking a release!

jeffreyguenther committed:
First pass on incoporating George's feedback

jeffreyguenther committed:
Merge branch 'master' into activestorage-guide

jeffreyguenther committed:
Adds Javan's event example

javan commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,397 @@
+
+Some services are supported by community maintained gems:
+
+* [OpenStack](https://github.com/jeffreyguenther/activestorage-openstack)
```
Do we have other lists like this in the Rails guides? I worry that it adds an unnecessary review burden going forward unless our policy is to add all submitted 3rd party libs.

I appreciate that you created this gem though! ❤️

jeffreyguenther replied:
I believe there are similar lists in Active Job.

https://github.com/rails/rails/blob/master/guides/source/active_job_basics.md#L163-L167

This is why I felt comfortable adding it.

jeffreyguenther replied:
I'm game to remove the section if the powers that be say that's what we should. If guides are only updated on a per release basis, I think that's probably best.

jeffreyguenther committed:
Change info to note

jeffreyguenther commented:
Ready for final review.

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,534 @@
+      params.require(:message).permit(:title, :content, images: [])
+    end
+end
+```
````
We can afford to expand on this, given that it’s the central feature of Active Storage:

> # Attach files to records
>
> ## `has_one_attached`
>
> The `has_one_attached` macro sets up a one-to-one mapping between records and files. Each record can have one file attached to it.
>
> For example, suppose your application has a `User` model. If you want each user to have an avatar, define the `User` model like this:
>
> ```ruby
> class User < ApplicationRecord
>   has_one_attached :avatar
> end
> ```
>
> You can create a user with an avatar:
>
> ```ruby
> class SignupController < ApplicationController
>   def create
>     user = Users.create!(user_params)
>     session[:user_id] = user.id
>     redirect_to root_path
>   end
>
>   private
>     def user_params
>       params.require(:user).permit(:email_address, :password, :avatar)
>     end
> end
> ```
>
> Call `avatar.attach` to attach an avatar to an existing user:
>
> ```ruby
> Current.user.avatar.attach(params[:avatar])
> ```
>
> Call `avatar.attached?` to determine whether a particular user has an avatar:
>
> ```ruby
> Current.user.avatar.attached?
> ```
>
> ## `has_many_attached`
>
> The `has_many_attached` macro sets up a one-to-many relationship between records and files. Each record can have many files attached to it.
>
> For example, suppose your application has a `Message` model. If you want each message to have many images, define the `Message` model like this:
>
> ```ruby
> class Message < ApplicationRecord
>   has_many_attached :images
> end
> ```
>
> You can create a message with images:
>
> ```ruby
> class MessagesController < ApplicationController
>   def create
>     message = Message.create!(message_params)
>     redirect_to message
>   end
>
>   private
>     def message_params
>       params.require(:message).permit(:title, :content, images: [])
>     end
> end
> ```
>
> Call `images.attach` to add new images to an existing message:
>
> ```ruby
> @message.images.attach(params[:images])
> ```
>
> Call `images.attached?` to determine whether a particular message has any images:
>
> ```ruby
> @message.images.attached?
> ```

jeffreyguenther commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,534 @@
+
+``` ruby
+gem 'mini_magick'
+```
````
Ya, I followed that, but it didn't feel right to me. 

We say:

> Continue reading for more information on the built-in service adapters (e.g.
`Disk` and `S3`) and the configuration they require.

As a reader, this led me to think that I would be receiving info about the different service types right away. It flowed better, to me, to have it below the mirrored service description.

I'll put it back the way you had it.

georgeclaghorn replied:
I see what you mean. Maybe we should move this information into the section on transforming images.

jeffreyguenther committed:
Incorporate latest round of feedback

pacMakaveli commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,544 @@
+config.active_storage.service = :local_test
+```
+
+Add Support Additional Cloud Service
````
I'm not a native English speaker, but this doesn't sound right. Maybe `Add Support for Additional Cloud Services` or `Add Support for an Additional Cloud Service`

jeffreyguenther replied:
Good catch. That's a typo!

pacMakaveli commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,544 @@
+Add Support Additional Cloud Service
+------------------------------------
+
+If you need to support a cloud service other these, you will need to implement
```
.. a cloud service other **than** these, you ..

pacMakaveli commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,544 @@
+@message.images.attach(params[:images])
+```
+
+Call `images.attached?`` to determine whether a particular message has any images:
````
Extra `

pacMakaveli commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,544 @@
+backups and migrations.
+
+Using Active Storage, an application can transform image uploads with
+[ImageMagick](https://www.imagemagick.org), generate image representations of
```
Which one is it? ImageMagick or MiniMagic or both? On line 329 you mention MiniMagick for creating variants.

jeffreyguenther replied:
MiniMagick is a ruby library wrapping the commandline utility ImageMagick. The args you pass to MiniMagick are the same as what you use on the commandline and the docs live over at ImageMagick explaining what they do. You're using ImageMagick via the MiniMagick.

pacMakaveli commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,544 @@
+end
+```
+
+If your system tests verify the deletion of a model with attachments and your
````
... with attachments and **you're** using ...

jeffreyguenther committed:
Merge branch 'master' into activestorage-guide

jeffreyguenther committed:
Apply edits and suggestions

georgeclaghorn commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,543 @@
+`Disk` and `S3`) and the configuration they require.
+
+### Disk Service
+Declare a Disk service in `config/storage.yml`:
```
Leave an extra line between the heading above and this paragraph.

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,543 @@
+  root: <%= Rails.root.join("storage") %>
+```
+
+### Amazon S3 Service
````
Leave an extra line between the heading above and this paragraph.

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,543 @@
+``` ruby
+gem "aws-sdk-s3", require: false
+```
+### Microsoft Azure Storage Service
````
Leave an extra line between the heading above and this paragraph.

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,543 @@
+gem "azure-storage", require: false
+```
+
+### Google Cloud Storage Service
````
Leave an extra line between the heading above and this paragraph.

georgeclaghorn commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,543 @@
+NOTE: Files are served from the primary service.
+
+Attach Files to a Model
+--------------------------
```
Match the number of dashes to the length of the heading.

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,543 @@
+```
+
+Remove File Attached to Model
+-------------------------------
````
Match the number of dashes to the length of the heading.

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,543 @@
+```
+
+Link to Attachments
+----------------------
````
Match the number of dashes to the length of the heading.

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,543 @@
+rails_blob_path(user.avatar, disposition: "attachment")
+```
+
+Create Variations of Attached Image
````
Let’s title this section “Transform Images.“

georgeclaghorn commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,543 @@
+-----------------------------------
+
+Sometimes your application requires images in a different format than
+what was uploaded. To create variation of the image, call `variant` on the Blob.
```
> To create a variation of an image…

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,543 @@
+<%= image_tag user.avatar.variant(resize: "100x100") %>
+```
+
+Create Image Previews of Attachments
````
> Preview Non-Image Files

georgeclaghorn commented on guides/source/active_storage_overview.md:
```diff
@@ -0,0 +1,543 @@
+Create Image Previews of Attachments
+------------------------------------
+Previews can be generated from some non-image formats. ActiveStorage supports
+Previewers for videos and PDFs.
```
Leave an extra line between the heading above and this paragraph.

Previewers are an implementation detail that I’d prefer not to expose in a guide intended for newcomers:

> Some non-image files can be previewed: that is, they can be presented as images. For example, a video file can be previewed by extracting its first frame. Out of the box, Active Storage supports previewing videos and PDF documents.

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,543 @@
+config.active_storage.service = :local_test
+```
+
+Add Support for Additional Cloud Services
````
> Support Additional Storage Services

georgeclaghorn commented on guides/source/active_storage_overview.md:
````diff
@@ -0,0 +1,543 @@
+```
+
+Add Support for Additional Cloud Services
+------------------------------------
````
Match the number of dashes to the length of the heading.

georgeclaghorn reviewed (changes requested):
This is my last round of review. Thanks so much for your work on this!

jeffreyguenther committed:
Merge branch 'master' into activestorage-guide

jeffreyguenther committed:
Formatting updates

jeffreyguenther committed:
Tweak whitespace

jeffreyguenther commented:
Here you go!
